src/integrationTest/java/org/opensearch/security/SecurityConfigurationBootstrapTests.java

```diff
@@ -0,0 +1,102 @@
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map;
+
+import com.carrotsearch.randomizedtesting.annotations.ThreadLeakScope;
+import org.apache.commons.io.FileUtils;
+import org.awaitility.Awaitility;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import org.opensearch.client.Client;
+import org.opensearch.security.securityconf.impl.CType;
+import org.opensearch.security.support.ConfigConstants;
+import org.opensearch.test.framework.TestSecurityConfig.User;
+import org.opensearch.test.framework.cluster.ClusterManager;
+import org.opensearch.test.framework.cluster.ContextHeaderDecoratorClient;
+import org.opensearch.test.framework.cluster.LocalCluster;
+import org.opensearch.test.framework.cluster.TestRestClient;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_ALLOW_DEFAULT_INIT_SECURITYINDEX;
+import static org.opensearch.security.support.ConfigConstants.SECURITY_RESTAPI_ROLES_ENABLED;
+import static org.opensearch.test.framework.TestSecurityConfig.Role.ALL_ACCESS;
+
+@RunWith(com.carrotsearch.randomizedtesting.RandomizedRunner.class)
+@ThreadLeakScope(ThreadLeakScope.Scope.NONE)
+public class SecurityConfigurationBootstrapTests {
+
+    private final static Path configurationFolder = ConfigurationFiles.createConfigurationDirectory();
+
+    private static final User USER_ADMIN = new User("admin").roles(ALL_ACCESS);
+
+    @ClassRule
+    public static LocalCluster cluster = new LocalCluster.Builder().clusterManager(ClusterManager.THREE_CLUSTER_MANAGERS)
+        .loadConfigurationIntoIndex(false)
+        .defaultConfigurationInitDirectory(configurationFolder.toString())
+        .nodeSettings(
+            Map.of(
+                SECURITY_RESTAPI_ROLES_ENABLED,
+                List.of("user_" + USER_ADMIN.getName() + "__" + ALL_ACCESS.getName()),
+                SECURITY_ALLOW_DEFAULT_INIT_SECURITYINDEX,
+                true
+            )
+        )
+        .build();
+
+    @BeforeClass
+    public static void initData() {
+
+        Client client = new ContextHeaderDecoratorClient(
+            cluster.getInternalNodeClient(),
+            Map.of(ConfigConstants.OPENDISTRO_SECURITY_CONF_REQUEST_HEADER, "true")
+        );
+        Runnable r = new Runnable() {
+            public void run() {
+                long t = System.currentTimeMillis();
+                long end = t + 10000;
+                while (System.currentTimeMillis() < end) {
+                    cluster.triggerConfigurationReloadForSingleCType(client, CType.CONFIG, true);
+                    try {
+                        Thread.sleep(50);
+                    } catch (InterruptedException e) {
+                        break;
+                    }
+                }
+            }
+        };
+
+        new Thread(r).start();
```

Comment:
Its to simulate running TransportConfigUpdateAction's while a node is still bootstrapping. This was how I was able to make an automated test for this scenario where a node can be bootstrapped with partial security config.

This runs a bunch of requests to trigger updating the config while the node is bootstrapping and ensures that the node finishes loading the entire security config instead of only the `config` portion of the securityconfig.

Reply:
I'm concerns this will introduce random failures as implemented - lets figure out how to better control the state for the test.

Ideally a test would start, setup the cluster in the invalid way, you'd be able to make sure the configuration was invalid, then you'd do something that would alter the state to make it valid, then repeat same check as before but get back a valid response.

Reply:
This is tricky to test since the requests need to be occurring during bootstrap while the bgThread is still running. I don't think there is a way to "pause" the bgThread in order to run a request and then resume to complete initialization. I can potentially look at https://github.com/opensearch-project/security/blob/main/src/test/java/org/opensearch/security/SlowIntegrationTests.java#L224-L258 as an example to delay initialization, but I think the bgThead will finish if it fails to create the security index.

Reply:
For your consideration, it might difficult to test these scenarios in an end to end test case.  There are two alternative paths you might want to consider, alter the test setup such as using several concrete components with mocks to stitch their interactions together, or by abstracting how the background process is triggered to be more testable in isolation.

Reply:
@peternied This is refactored now and I introduced a new setting to delay initialization to help with testing this.

I ran into more trouble writing an automated test for this because I kept hitting this [error message](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/DynamicConfigFactory.java#L254). This error message would trigger when the node booting up received a TransportConfigUpdateAction to reload the CONFIG ctype when the config cache was still empty. When receiving the action it would query for the `config` portion of the security index and populate the corresponding entry in the cache while other entries, like `role`, were empty. When role was empty in the cache, this [line](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/DynamicConfigFactory.java#L171) would return a cache miss and the call to `roles.add(staticRoles)` fails because it has an empty roles config. 

In order to simulate the issue on ConfigModelV7, I trigger the ConfigUpdateAction with 4 cTypes (actiongroups, config, roles, tenants) and don't pass others in order for it to pass the validation logic for ConfigModelV7.